Merge repair and rearm anim traits into WithResupplyAnimation

reaperrr opened the pull request:
Another (and definitely the last) split from #14357.

This is the safest approach to avoid animation conflicts/visual glitches when the 'host' is responsible for both resupply types.
The new `WithResupplyAnimation` trait will simply play a looping animation as long as the actor is resupplying in any form.

As a byproduct, this makes the repair/reload animations look more like in the originals and less erratic.
We don't have a `WithRearmOverlay` yet, so I'll leave a refactor of `WithRepairOverlay` for later.

reaperrr committed:
Merge repair and rearm anim traits into WithResupplyAnimation

This is the safest approach to avoid conflicts/visual glitches when the host is responsible for both resupply types.
The new trait will simply play a looping animation as long as the actor is resupplying in any form.

reaperrr commented:
@Mailaender regarding your comment on IRC, I'll add some context:

This was split from a larger PR.

Basically, the ultimate goal of this PR, #14357 which it was split from and the PR(s) that will follow, is to allow actors to be reloaded and repaired _at the same time_ (right now on bleed, actors that need both will be rearmed first, then repaired, as both are separate activities that are queued in `Repairable`).

The problem is that on bleed, the reload/repair animation (of the building) is played once on every repair/reload step, instead of just looping continuously. 
So when repairing and rearming happen in parallel once the refactoring is done, depending on their interval it's possible that, for example, a repair step happens just a tick after a rearm step, resulting in restarting the service depot animation before the last one finished.
This is also already an issue with rearming or repairing when the resupply interval is shorter than the animation, as in that case it restarts prematurely as well.

This PR simply makes the resupply animations of helipads, service depots and so on loop at consistent speed as long as the building is either repairing or rearming (or both) some actor.

By the way, this also makes the animations look more like in the original, where the animation speed wasn't tied to the repair/reload speed, either, making it look less hectic.

Mailaender commented:
If a vehicle gets killed while repairing/rearming the animation continues:

![image](https://user-images.githubusercontent.com/756669/45595987-1063df80-b9b5-11e8-8066-1430033a2367.png)

I don't know if that is a regression or already existing.

Mailaender reviewed (dismissed):
Rearming and repairing still works and the animation plays.

reaperrr commented:
> If a vehicle gets killed while repairing/rearming the animation continues

Should be an easy fix, if that's the case I might as well include it here.

reaperrr commented:
Rebased onto #15622 and fixed the above issue.

reaperrr commented:
Rebased onto updated #15622 and fixed that ResupplyAircraft only notified INotifyRepair (unconditionally).
It now checks which of the activities was running on disposal and then notifies the correct trait(s) accordingly.

reaperrr commented:
Rebased and updated again.
The update only affects `Rearm` activity though, it now properly gets the host passed like `Repair` instead of relying on the old ActorsAt hacks.

According to my own tests everything still works as it should, but I guess there have been too many changes since @Mailaender's review to keep his +1.

abcdefg30 reviewed (dismissed):
Looks fine otherwise and ingame. :+1: 
Do we want to move the update rule into a folder like in #14664?

reaperrr commented on OpenRA.Mods.Common/Traits/Render/WithResupplyAnimation.cs:
```diff
@@ -0,0 +1,113 @@
+		public readonly string Body = "body";
+
+		[Desc("Events leading to the animation getting played. Possible values currently are: Rearm, Repair.")]
+		public readonly ResupplyType PlayAnimationOn = ResupplyType.Rearm | ResupplyType.Repair;
```
Not sure, tbh. Removed, we can always add it back later if needed, afterall.

reaperrr commented:
Updated.

> Do we want to move the update rule into a folder like in #14664?

Done.

reaperrr commented:
Rebased.

abcdefg30 commented:
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/d5913ed49be02ef26ec6a772051f6b87baade90e...b8a3afc425a2588092f098e49b9e0f251723972e)
